api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto

```diff
@@ -183,44 +183,151 @@ message JwtHeader {
   string value_prefix = 2;
 }
 
+// This message specifies a Jwt requirement.
+message JwtRequirement {
+  // Specify the required issuer for the JwtProvider.
+  string issuer = 1;
+
+  // This audiences field overrides the one in JwtProvider if not empty.
+  repeated string audiences = 2;
+}
+
+// This message specifies a per route based Jwt config.
+// Route.per_filter_config for jwt_authn filter uses this message.
+message PerRouteConfig {
+  // Specify list of required JwtProviders. If list is empty,  Jwt is NOT required.
+  // If any of required JwtProviders is missing, or its verification fails, the request will
+  // be rejected.
+  // If a Jwt is not required, but it is in the providers list, it will be verified.
+  // But it will not fail the request if its verification fails.
+  repeated JwtRequirement requires = 1;
+}
+
+// This message specifies a Jwt requirement for a specific Route condition.
+// Example 1:
+//
+// .. code-block:: yaml
+//
+//    - match:
+//         prefix: "/healthz"
+//
+// In above example, There is no "requires" field for /healthz prefix match,
+// it means that requrests matched the path prefix don't require jwt authentication.
+//
+// Example 2:
+//
+// .. code-block:: yaml
+//
+//    - match:
+//         prefix: "/"
+//      requires:
+//         - issuer: jwt_issuer1
+//         - issuer: jwt_issuer2
+//           audiences: [audience1, audience2]
+//
+// In above example, all requests (all requests matched "/" path prefix),
+// require jwt authentication from [jwt_issuer1, jwt_issuer2] and jwt_issuer2 has audiences
+// override defined.
+message RequirementRule {
+  // The route matching parameter. Only when the match is satisfied, the "requires" field will
+  // apply.
+  //
+  // For example: following match will match all requests.
+  //
+  // .. code-block:: yaml
+  //
+  //    match:
+  //       prefix: "/"
+  //
+  envoy.api.v2.route.RouteMatch match = 1 [(validate.rules).message.required = true];
+
+  // Specify list of required JwtProviders. If list is empty,  Jwt is NOT required.
+  // If any of required JwtProviders is missing, or its verification fails, the request will
+  // be rejected.
+  // If a Jwt is not required, but it is in the providers list, it will be verified.
+  // But it will not fail the request if its verification fails.
+  repeated JwtRequirement requires = 2;
```

Comment:
If I'm reading the comment correctly, this means we're AND-ing all the requirements? Is that more useful than OR-ing the requirements? I'd think that a typical request has one JWT and would thus only match one issuer, so it's unlikely for AND-ed requirements to be satisfied unless they all have the same issuer and only differ in audience + the JWT has multiple audiences defined.  Having OR-ed requirements means the route can accept tokens from multiple issuers, which seems like a more common use case. Thoughts?

Reply:
I think AND-ing the requirement will be more useful.   If you ever want to verify two JWT from different issuers,  you most likely want both of them to pass, not just one of them.   For example,  Google cloud supports IAP and end-user token,  some requests want both to pass, it will be: [iap,  end_user_jwt],  some only want iap,  it will be [iap],   even the requests may carry end_user_jwt,  but it is not required.  It is very hard for me to imaging a request only wants either iap or end_user_jwt.

Reply:
I see, so the request carries multiple JWTs in different headers/params? If so AND-ing makes sense, because otherwise there isn't a good way to specify it.

Reply:
@qiwzhang we do OR-ed requirements in Google Cloud, https://github.com/googleapis/googleapis/blob/master/google/api/auth.proto#L61 . It is useful to support AND-ing when you have multiple JWTs. However, more common cases we've seen are: your service serve both internal service-to-service traffic and ingress traffic, then you need to specify `JWT signed by a service account` OR `End-user JWT issued by A` OR `End-user JWT issued by B`.

Reply:
Agree with @lizan , I'd prefer OR too, at least at the top level. Just seems more common to allow multiple issuers to authenticate a user.

However, I could see multi-JWT requests being used, so perhaps we can come up with something that satisfies both? I'd rather start off with a slightly more verbose config vs change it down the road.

A few misc notes that come to mind for multi-JWT configuration:
- We could do a "2-d" list, where the top level iterates over OR-ed requirements and the inner is AND-ed. Or maybe use a union or oneof?
- We'd probably want a way to map headers to requirements explicitly, otherwise the implementation would have to try each of the N^2 pairs. So either the requirements need to specify the `from_headers` or `from_params`, or the requirements need ids and there needs to be some mapping elsewhere.

Reply:
Let's not make things too complicate, we might be fine with OR here, but if there are multiple JWTs, validate all of them, and let authZ (RBAC) to handle the complex AND/OR rules.